File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramListDiffCallback.java
package org.tvheadend.tvhclient.ui.features.programs;

import org.tvheadend.tvhclient.domain.entity.Program;

import java.util.List;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

class ProgramListDiffCallback extends DiffUtil.Callback {
    private final List<Program> oldList;
    private final List<Program> newList;

    ProgramListDiffCallback(List<Program> oldList, List<Program> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList != null ? oldList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return newList != null ? newList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return newList.get(newItemPosition).getEventId() == oldList.get(oldItemPosition).getEventId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return newList.get(newItemPosition).equals(oldList.get(oldItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        // One can return particular field for changed item.
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramListDiffCallback.kt
package org.tvheadend.tvhclient.ui.features.programs

import androidx.recyclerview.widget.DiffUtil
import org.tvheadend.tvhclient.data.entity.ProgramInterface

internal class ProgramListDiffCallback(private val oldList: List<ProgramInterface>, private val newList: List<ProgramInterface>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return newList[newItemPosition].eventId == oldList[oldItemPosition].eventId
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return newList[newItemPosition] == oldList[oldItemPosition]
    }
}
